plugins/modem-manager/fu-plugin-modem-manager.c

```diff
@@ -0,0 +1,472 @@
+	g_free (info->inhibited_uid);
+	g_free (info->physical_id);
+	g_free (info->vendor);
+	g_free (info->name);
+	g_free (info->version);
+	g_strfreev (info->guids);
+	g_free (info->detach_fastboot_at);
+	g_free (info);
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (FuPluginMmInhibitedDevice, fu_plugin_mm_inhibited_device_free);
+
+static void
+fu_plugin_mm_udev_device_removed (FuPlugin *plugin)
+{
+	FuPluginData *priv = fu_plugin_get_data (plugin);
+	FuMmDevice *dev;
+
+	if (priv->inhibited == NULL)
+		return;
+
+	dev = fu_plugin_cache_lookup (plugin, priv->inhibited->physical_id);
+	if (dev == NULL)
+		return;
+
+	/* once the first port is gone, consider device is gone */
+	fu_plugin_cache_remove (plugin, priv->inhibited->physical_id);
+	fu_plugin_device_remove (plugin, FU_DEVICE (dev));
+
+	/* no need to wait for more ports, cancel that right away */
+	if (priv->inhibited->udev_timeout_id != 0) {
+		g_source_remove (priv->inhibited->udev_timeout_id);
+		priv->inhibited->udev_timeout_id = 0;
+	}
+}
+
+static void
+fu_plugin_mm_uninhibit_device (FuPlugin *plugin)
+{
+	FuPluginData *priv = fu_plugin_get_data (plugin);
+	g_autoptr(FuPluginMmInhibitedDevice) info = NULL;
+
+	/* get the device removed from the plugin cache before uninhibiting */
+	fu_plugin_mm_udev_device_removed (plugin);
+
+	info = g_steal_pointer (&priv->inhibited);
+	if ((priv->manager != NULL) && (info != NULL)) {
+		g_debug ("uninhibit modemmanager device with uid %s", info->inhibited_uid);
+		mm_manager_uninhibit_device_sync (priv->manager, info->inhibited_uid, NULL, NULL);
+	}
+}
+
+static gboolean
+fu_plugin_mm_udev_device_ports_timeout (gpointer user_data)
+{
+	FuPlugin *plugin = user_data;
+	FuPluginData *priv = fu_plugin_get_data (plugin);
+	FuMmDevice *dev;
+	g_autoptr(GError) error = NULL;
+
+	g_assert (priv->inhibited != NULL);
+	priv->inhibited->udev_timeout_id = 0;
+
+	dev = fu_plugin_cache_lookup (plugin, priv->inhibited->physical_id);
+
+	if (dev != NULL) {
+		if (!fu_device_probe (FU_DEVICE (dev), &error)) {
+			g_debug ("failed to probe MM device: %s", error->message);
+		} else {
+			fu_plugin_device_add (plugin, FU_DEVICE (dev));
+		}
+	}
+
+	return G_SOURCE_REMOVE;
+}
+
+static void
+fu_plugin_mm_udev_device_ports_timeout_reset (FuPlugin *plugin)
+{
+	FuPluginData *priv = fu_plugin_get_data (plugin);
+
+	g_assert (priv->inhibited != NULL);
+	if (priv->inhibited->udev_timeout_id != 0)
+		g_source_remove (priv->inhibited->udev_timeout_id);
+	priv->inhibited->udev_timeout_id = g_timeout_add_seconds (FU_MM_UDEV_DEVICE_PORTS_TIMEOUT,
+								  fu_plugin_mm_udev_device_ports_timeout,
+								  plugin);
+}
+
+static void
+fu_plugin_mm_udev_device_port_added (FuPlugin		*plugin,
+				     const gchar	*subsystem,
+				     const gchar	*path,
+				     gint		 ifnum)
+{
+	FuPluginData *priv = fu_plugin_get_data (plugin);
+	FuMmDevice *existing;
+	g_autoptr(FuMmDevice) dev = NULL;
+	g_autoptr(GError) error = NULL;
+
+	g_assert (priv->inhibited != NULL);
```

Comment:
Can we make this a g_warning or g_critical instead please? If the daemon asserts I get bugs created from abrt and I look bad on various charts. Thanks.

Reply:
:D

Reply:
Would you like me to remove all asserts? Or just this one? Because I usually use asserts when the logic should really not fail in the implementation. I don't think that converting that into a warning makes much sense, because it would segfault later on anyway if it ever fails.

Reply:
I don't like g_assert() -- I'd much rather it be a g_return_val_if_fail() or g_critical and return.

Reply:
ok, I will convert the asserts to g_return_if_fails then I think